src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -1172,6 +1179,247 @@ void SolverInterfaceImpl::writeScalarData(
                 "Please make sure you only use the results from calls to setMeshVertex/Vertices().",
                 context.getDataName(), valueIndex);
   values[valueIndex] = value;
+
+  PRECICE_DEBUG("Written Scalar Value = {}", value);
+}
+
+void SolverInterfaceImpl::writeScalarGradientData(
+    int           dataID,
+    int           valueIndex,
+    const double *value)
+{
+
+  PRECICE_TRACE(dataID, valueIndex);
+  PRECICE_CHECK(_state != State::Finalized, "writeScalarGradientData(...) cannot be called before finalize().")
+  PRECICE_REQUIRE_DATA_WRITE(dataID);
+  PRECICE_DEBUG("gradient value = {}", Eigen::Map<const Eigen::VectorXd>(value, _dimensions).format(utils::eigenio::debug()));
+
+  WriteDataContext &context = _accessor->writeDataContext(dataID);
+  PRECICE_ASSERT(context.providedData() != nullptr);
+  mesh::Data &data = *context.providedData();
+
+  //Check if data has been initialized to include gradient data
+  PRECICE_CHECK(data.hasGradient(), "Data \"{}\" has no gradient values available! Please turn on the gradient attribute in the configuration file.", data.getName())
+
+  // Size of the gradient data input : must be spaceDimensions * dataDimensions -> here spaceDimensions (since for scalar: dataDimensions = 1)
+  PRECICE_CHECK(data.getSpacialDimensions() == _dimensions,
+                "The spatial dimensions of the data object and of the mesh don't match. Please make sure to initialize \"{0}\" space dimensions for the scalar data type \"{1}\".",
+                _dimensions, context.getDataName());
+  PRECICE_VALIDATE_DATA(value, _dimensions);
+
+  // Gets the gradientvalues matrix corresponding to the dataID
+  auto &     gradientValues = data.gradientValues();
+  const auto vertexCount    = gradientValues.cols() / context.getDataDimensions();
+
+  //Check if the index and dimensions are valid
+  PRECICE_CHECK(valueIndex >= -1,
+                "Invalid value index ({}) when writing gradient scalar data. Value index must be >= 0. "
```

Comment:
I implemented this consistently as in the ``writeData``-methods. But you are right it does make more sense. Should I change it for all the methods or just the gradient methods ?

Reply:
I implemented this consistently with the ``writeData``-methods. I think the reason we don't call it ``vertexIndex`` is because it's not the absolute index of the vertex in the whole mesh but rather just the index in the current block set of data input.
I am not sure but I can change it nevertheless, if you think it's more fitting.